test/infrastructure/kubemark/controllers/kubemarkmachine_controller.go

```diff
@@ -0,0 +1,451 @@
+	}
+	machinePatchHelper, err := patch.NewHelper(machine, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to init patch helper: %w", err)
+	}
+	defer func() {
+		if err := machinePatchHelper.Patch(context.TODO(), machine); err != nil {
+			if !apierrors.IsNotFound(err) {
+				logger.Error(err, "failed to patch machine")
+			}
+		}
+	}()
+
+	logger = logger.WithValues("machine", machine.Name)
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		logger.Info("Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, nil
+	}
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	restConfig, _, err := getRemoteCluster(ctx, logger, r.Client, cluster)
+	if err != nil {
+		logger.Error(err, "error getting remote cluster")
+		return ctrl.Result{}, err
+	}
+
+	if !cluster.Status.InfrastructureReady {
+		logger.Info("Cluster infrastructure is not ready yet")
+		return ctrl.Result{}, nil
+	}
+	if machine.Spec.Bootstrap.DataSecretName == nil {
+		logger.Info("Bootstrap data secret reference is not yet available")
+		return ctrl.Result{}, nil
+	}
+
+	var kubeadmConfig bootstrapv1.KubeadmConfig
+	if err := r.Get(context.TODO(), types.NamespacedName{
+		Name:      machine.Spec.Bootstrap.ConfigRef.Name,
+		Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace,
+	}, &kubeadmConfig); err != nil {
+		logger.Error(err, "error getting bootstrap config")
+		return ctrl.Result{}, err
+	}
+
+	cfg, err := RetrieveValidatedConfigInfo(kubeadmConfig.Spec.JoinConfiguration)
+	if err != nil {
+		logger.Error(err, "error validating token")
+		return ctrl.Result{}, err
+	}
+
+	clusterinfo := cfg.Clusters[""]
+	cfg = CreateWithToken(
+		clusterinfo.Server,
+		DefaultClusterName,
+		TokenUser,
+		clusterinfo.CertificateAuthorityData,
+		kubeadmConfig.Spec.JoinConfiguration.Discovery.BootstrapToken.Token,
+	)
+	certificateStore := &capkcert.MemoryStore{}
+
+	newClientFn := func(current *tls.Certificate) (clientset.Interface, error) {
+		client, err := clientset.NewForConfig(restConfig)
+		if err != nil {
+			logger.Error(err, "error creating clientset")
+			return nil, err
+		}
+		return client, nil
+	}
+	mgr, err := certificate.NewManager(&certificate.Config{
+		BootstrapCertificatePEM: cfg.AuthInfos[TokenUser].ClientCertificateData,
+		BootstrapKeyPEM:         cfg.AuthInfos[TokenUser].ClientKeyData,
+		CertificateStore:        certificateStore,
+		Template: &x509.CertificateRequest{
+			Subject: pkix.Name{
+				CommonName:   fmt.Sprintf("system:node:%s", kubemarkMachine.Name),
+				Organization: []string{"system:nodes"},
+			},
+		},
+		Usages: []certificates.KeyUsage{
+			certificates.UsageDigitalSignature,
+			certificates.UsageKeyEncipherment,
+			certificates.UsageClientAuth,
+		},
+		ClientsetFn: newClientFn,
+	})
+	if err != nil {
+		logger.Error(err, "error creating cert manager")
+		return ctrl.Result{}, err
+	}
+
+	mgr.Start()
```

Comment:
We already have access to the cluster CA (through the secrets), I suspect quite a bit of this could be simplified by using that CA rather than using the bootstrap token to request a certificate to be signed by the cluster CA

That said, it might be good to keep this workflow if we intend to support control plane providers other than Kubeadm.

Reply:
Yeah, I reverse-engineered this by figuring out how both the `kubeadm join` and kubelet bootstrap phases converted the token into a full-fledge kubeconfig. I am sure it can be made simpler!

Reply:
If that's the case, then I'd say you can likely just rip out a good chunk of this by snagging the CA secret and creating the node cert/kubeconfig directly. We can always add additional support later if need be.